third_party/move/move-compiler-v2/src/file_format_generator/function_generator.rs

```diff
@@ -653,6 +1160,320 @@ impl<'a> FunctionGenerator<'a> {
+                    self.abstract_push_result(ctx, dest);
+                    found = true;
+                    break;
+                } else if let Some(cur_variants) = cur_variants_opt {
+                    // handle borrow from an enum
+                    assert!(variants.is_some());
+                    let cur_variants_set = cur_variants.iter().collect::<BTreeSet<_>>();
+                    let variants_set = variants.as_ref().unwrap().iter().collect::<BTreeSet<_>>();
+                    // case 1: exact match
+                    if cur_variants_set == variants_set {
+                        self.abstract_push_args(ctx, source, None);
+                        // need to freeze mutable ref if the target is an immutable reference
+                        if is_mut_src && !is_mut_dest {
+                            self.emit(FF::Bytecode::FreezeRef);
+                        }
+                        if inst.is_empty() {
+                            self.emit(FF::Bytecode::Call(*borrow_field_handle_index));
+                        } else {
+                            let function_instantiation_index =
+                                self.r#gen.function_instantiation_index_from_handle_index(
+                                    *borrow_field_handle_index,
+                                    &ctx.fun_ctx.module,
+                                    &fun_ctx.loc,
+                                    inst.to_vec(),
+                                );
+                            self.emit(FF::Bytecode::CallGeneric(function_instantiation_index));
+                        }
+                        self.abstract_pop_n(ctx, source.len());
+                        self.abstract_push_result(ctx, dest);
+                        found = true;
+                        break;
+                    } else if variants_set.is_subset(&cur_variants_set) {
+                        // case 2: variants in the borrow field are a subset of the variants in the api
+                        let intersected_variants = variants_set
+                            .intersection(&cur_variants_set)
+                            .collect::<Vec<_>>();
+                        let test_variant_api_map = self.r#gen.struct_api_test_variant_index(
+                            &ctx.fun_ctx.module,
+                            &fun_ctx.loc,
+                            struct_env,
+                        );
+                        // before calling the test variant function, we need to flush the stack
+                        self.abstract_flush_stack_before(ctx, 0);
+                        // the base code offset to compute the target offset for each test variant
+                        let code_offset_for_test_variant = self.code.len();
+                        let variants_size = intersected_variants.len();
+                        // offset count for each test variant:
+                        // 0. copy source
```

Comment:
I think we could do freezeref only once (if needed), and then copy the frozen ref for any later uses instead of copying the source and freezing it for every variant?